### ISSUE TITLE
Harbor full scan make almost all images in pending state preventing pulls

### ISSUE THREAD

**Issue by hamdikh** [goharbor/harbor#19385]
Hello, we have a few questions related to how harbor full scanning.
We have more than 57K images and the global scan is taking more than 12 hours to complete.
In our configuration, we've 450 job service workers and we're having 60 parallel scans.

PS: we started experiencing these problems after the v2.7.3 upgrade, we've moved from v2.6.4 to v2.7.3 to resolve database CPU consumption issues. 
We're seeing a jobservice impact on how the queuing system works as the full scan puts the images in a pending state.
As we've also configured the prevent images with vulns from being pulled this blocks all our image pulls.

We would like to know if there are any optimizations that can be done from our side. 
Any help will be much appreciated.

Harbor v2.7.3
Kubernetes version: 1.25.11

**Comment by hamdikh** [comment]
I provided a little bit more context to reproduce, I hope this helps

**Comment by jthin** [comment]
Hello,

To add additional information about Hamdi's issue:

- During a full scan, when an image is in the *pending* queue, the last scan report seems to be deleted, and the image is left waiting to be scanned.
- Since we are preventing project teams from pulling images with a severity level, the image must be scanned to obtain a Trivy scan result, even if the image has already been scanned during the initial push or has been manually scanned within the last few days/hours.
- If there are many images in the *pending* queue, project teams may have to wait for hours before they can pull the image.

At the moment, we do not have the option to perform a weekly full scan, and we hope that no project team will be affected by this issue.

**Comment by abenabid** [comment]
Hello,
I've noticed that when the scan task is added to the queue, the previous scan report is removed from the database.
IMO, the previous scan report is still legitimate, so it should bet kept in the database until the scanner finishes the scan **successfully**.
Can this solve the image pull issue ?

**Comment by sizowie** [comment]
Please do not close.

**Comment by hamdikh** [comment]
Issue not stale please don't close

**Comment by jthin** [comment]
Not stale

**Comment by bupd** [comment]
Yes, this behavior exists. One way to improve it would be not deleting previous scan results, and instead serving images based on the last known scan results until new ones are available.


Day 1: You scan an image => it only has low and medium vulnerabilities. Since prevention is set to block only critical vulnerabilities, the image passes and is allowed to pull.

Day 2: You rescan the same image => now a critical vulnerability is discovered.

This raises two possible approaches:

1. Serve images until the new scan completes (using previous scan results).
2. Block image pulls until the new scan completes (no serving until updated results are in).

I recommend Option 1, because most images will retain similar vulnerability levels between scans.
> If the image was already deployed previously, it makes sense to continue serving it until the scanner confirms new critical vulnerabilities.

**Comment by hamdikh** [comment]
The actual issue is in the scanning queue. When Day 2 comes, the scanner blocks all image pulls not just the ones pending a new scan. That means even unscanned images get blocked, effectively stopping all pulls from the registry.

So the problem isn’t only about whether to serve the last known scan results vs. wait for new ones, but rather that the queue itself causes a full registry block once rescanning kicks in.
